perfkitbenchmarker/openstack/os_virtual_machine.py

```diff
@@ -0,0 +1,190 @@
+# Copyright 2015 Google Inc. All rights reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#   http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import logging
+import random
+import string
+import time
+
+from perfkitbenchmarker import virtual_machine, linux_virtual_machine
+from perfkitbenchmarker import flags
+from perfkitbenchmarker import vm_util
+from perfkitbenchmarker.openstack import os_disk
+from perfkitbenchmarker.openstack import utils as os_utils
+
+UBUNTU_IMAGE = 'ubuntu-14.04'
+
+FLAGS = flags.FLAGS
+
+flags.DEFINE_boolean('openstack_config_drive', False,
+                     'Add possibilities to get metadata from external drive')
+
+flags.DEFINE_boolean('openstack_boot_from_volume', False,
+                     'Boot from volume instead of an image')
+
+flags.DEFINE_integer('openstack_volume_size', 20,
+                     'Size of the volume (GB)')
+
+
+def id_generator(size=6, chars=string.ascii_uppercase + string.digits):
+    return ''.join(random.choice(chars) for _ in range(size))
+
+
+class OpenStackVirtualMachine(virtual_machine.BaseVirtualMachine):
+    """Object representing an OpenStack Virtual Machine"""
+
+    DEFAULT_MACHINE_TYPE = 'm1.small'
+    DEFAULT_ZONE = 'nova'
+    # Subclasses should override the default image.
+    DEFAULT_IMAGE = None
+
+    def __init__(self, vm_spec):
+        super(OpenStackVirtualMachine, self).__init__(vm_spec)
+        self.special_id = id_generator()
+        self.name = 'perfkit_vm_%d_%s' % (self.instance_number,
+                                          self.special_id)
+        self.key_name = 'perfkit_key_%d_%s' % (self.instance_number,
+                                               self.special_id)
+        self.client = os_utils.NovaClient()
+        self.id = -1
+        self.pk = -1
+        self.user_name = 'ubuntu'
+        self.wait_time = 15
```

Comment:
Changed initialization to None, and moved the value updates to WaitForBootCompletion with a comment added so that it's all in one place.